server/schedule/operator/builder.go

```diff
@@ -0,0 +1,596 @@
+		return best
+	} else if mb < mc {
+		best = stepPlan{}
+	}
+
+	// Brute force all possible leader combinations to find the best plan.
+	for _, leader1 := range b.currentPeers.IDs() {
+		if !b.allowLeader(b.currentPeers.Get(leader1)) {
+			continue
+		}
+		next.leader1 = leader1
+		for _, leader2 := range b.currentPeers.IDs() {
+			if b.allowLeader(b.currentPeers.Get(leader2)) {
+				next.leader2 = leader2
+				best = b.comparePlan(best, next)
+			}
+		}
+		if b.allowLeader(next.addOrPromote) {
+			next.leader2 = next.addOrPromote.GetStoreId()
+			best = b.comparePlan(best, next)
+		}
+	}
+	return best
+}
+
+func (b *Builder) planRemovePeer() stepPlan {
+	var best stepPlan
+	for _, i := range b.toRemove.IDs() {
+		r := b.toRemove.Get(i)
+		for _, leader := range b.currentPeers.IDs() {
+			if b.allowLeader(b.currentPeers.Get(leader)) {
+				best = b.comparePlan(best, stepPlan{remove: r, leader2: leader})
+			}
+		}
+	}
+	return best
+}
+
+func (b *Builder) planAddPeer() stepPlan {
+	var best stepPlan
+	for _, i := range b.toAdd.IDs() {
+		a := b.toAdd.Get(i)
+		for _, leader := range b.currentPeers.IDs() {
+			if b.allowLeader(b.currentPeers.Get(leader)) {
+				best = b.comparePlan(best, stepPlan{addOrPromote: a, leader1: leader})
+			}
+		}
+	}
+	return best
+}
+
+func (b *Builder) planPromotePeer() stepPlan {
+	for _, i := range b.toPromote.IDs() {
+		p := b.toPromote.Get(i)
+		return stepPlan{addOrPromote: p, leader1: b.currentLeader}
+	}
+	return stepPlan{}
+}
+
+// Pick the better plan from 2 candidates.
+func (b *Builder) comparePlan(best, next stepPlan) stepPlan {
+	if best.empty() {
+		return next
+	}
+	fs := []func(stepPlan) int{
+		b.avoidRemoveLeader,
+		b.preferUpStoreAsLeader,
+		b.preferOldPeerAsLeader,
+		b.preferLessLeaderTransfer1,
+		b.preferLessLeaderTransfer2,
+		b.preferTargetLeader,
+	}
+	for _, t := range fs {
+		if tb, tc := t(best), t(next); tb > tc {
+			return best
+		} else if tb < tc {
+			return next
+		}
+	}
+	return best
+}
+
+func (b *Builder) labelMatch(p stepPlan) int {
+	if p.empty() {
+		return 0
+	}
+	sx, sy := b.cluster.GetStore(p.addOrPromote.GetStoreId()), b.cluster.GetStore(p.remove.GetStoreId())
+	if sx == nil || sy == nil {
+		return 0
+	}
+	labels := b.cluster.GetLocationLabels()
+	for i, l := range labels {
+		if sx.GetLabelValue(l) != sy.GetLabelValue(l) {
+			return i
+		}
+	}
+	return len(labels)
+}
+
+func b2i(b bool) int { // revive:disable-line:flag-parameter
+	if b {
+		return 1
+	}
+	return 0
+}
+
+// If remove leader directly, tikv will need to wait for raft election timeout.
+func (b *Builder) avoidRemoveLeader(p stepPlan) int {
+	return b2i(p.remove == nil || p.leader2 != p.remove.GetStoreId())
+}
+
+// Avoid generating snapshots from offline stores.
+func (b *Builder) preferUpStoreAsLeader(p stepPlan) int {
+	if p.addOrPromote != nil && b.currentPeers.Get(p.addOrPromote.GetStoreId()) == nil {
+		store := b.cluster.GetStore(p.leader1)
```

Comment:
Do we need to check `leader2`, too?

Reply:
This rule is to avoid generating snapshot from offline stores and only leader1 generates snapshot.

Reply:
I see.